feat: bundle sparse package to integrate with windows context menu

deepak1556 opened the pull request:
Fixes https://github.com/microsoft/vscode/issues/127365

- [x] ESRP signing
- [x] Inno install/uninstall
- [x] Update Flow

deepak1556 commented on build/win32/code.iss:
```diff
@@ -1262,17 +1267,17 @@ Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\Applications\{#ExeBas
-Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\Drive\shell\{#RegValueName}"; ValueType: expandsz; ValueName: ""; ValueData: "{cm:OpenWithCodeContextMenu,{#ShellNameShort}}"; Tasks: addcontextmenufolders; Flags: uninsdeletekey
-Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\Drive\shell\{#RegValueName}"; ValueType: expandsz; ValueName: "Icon"; ValueData: "{app}\{#ExeBasename}.exe"; Tasks: addcontextmenufolders
-Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\Drive\shell\{#RegValueName}\command"; ValueType: expandsz; ValueName: ""; ValueData: """{app}\{#ExeBasename}.exe"" ""%V"""; Tasks: addcontextmenufolders
+Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\*\shell\{#RegValueName}"; ValueType: expandsz; ValueName: "Icon"; ValueData: "{app}\{#ExeBasename}.exe"; Tasks: addcontextmenufiles; OnlyBelowVersion: 10.0.22000
+Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\*\shell\{#RegValueName}\command"; ValueType: expandsz; ValueName: ""; ValueData: """{app}\{#ExeBasename}.exe"" ""%1"""; Tasks: addcontextmenufiles; OnlyBelowVersion: 10.0.22000
+Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\directory\shell\{#RegValueName}"; ValueType: expandsz; ValueName: ""; ValueData: "{cm:OpenWithCodeContextMenu,{#ShellNameShort}}"; Tasks: addcontextmenufolders; Flags: uninsdeletekey; OnlyBelowVersion: 10.0.22000
+Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\directory\shell\{#RegValueName}"; ValueType: expandsz; ValueName: "Icon"; ValueData: "{app}\{#ExeBasename}.exe"; Tasks: addcontextmenufolders; OnlyBelowVersion: 10.0.22000
+Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\directory\shell\{#RegValueName}\command"; ValueType: expandsz; ValueName: ""; ValueData: """{app}\{#ExeBasename}.exe"" ""%V"""; Tasks: addcontextmenufolders; OnlyBelowVersion: 10.0.22000
+Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\directory\background\shell\{#RegValueName}"; ValueType: expandsz; ValueName: ""; ValueData: "{cm:OpenWithCodeContextMenu,{#ShellNameShort}}"; Tasks: addcontextmenufolders; Flags: uninsdeletekey; OnlyBelowVersion: 10.0.22000
+Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\directory\background\shell\{#RegValueName}"; ValueType: expandsz; ValueName: "Icon"; ValueData: "{app}\{#ExeBasename}.exe"; Tasks: addcontextmenufolders; OnlyBelowVersion: 10.0.22000
+Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\directory\background\shell\{#RegValueName}\command"; ValueType: expandsz; ValueName: ""; ValueData: """{app}\{#ExeBasename}.exe"" ""%V"""; Tasks: addcontextmenufolders; OnlyBelowVersion: 10.0.22000
+Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\Drive\shell\{#RegValueName}"; ValueType: expandsz; ValueName: ""; ValueData: "{cm:OpenWithCodeContextMenu,{#ShellNameShort}}"; Tasks: addcontextmenufolders; Flags: uninsdeletekey; OnlyBelowVersion: 10.0.22000
+Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\Drive\shell\{#RegValueName}"; ValueType: expandsz; ValueName: "Icon"; ValueData: "{app}\{#ExeBasename}.exe"; Tasks: addcontextmenufolders; OnlyBelowVersion: 10.0.22000
+Root: {#SoftwareClassesRootKey}; Subkey: "Software\Classes\Drive\shell\{#RegValueName}\command"; ValueType: expandsz; ValueName: ""; ValueData: """{app}\{#ExeBasename}.exe"" ""%V"""; Tasks: addcontextmenufolders; OnlyBelowVersion: 10.0.22000
```
Avoid installing these registry keys for windows 11 or later. The default key for title is left in place since there is no efficient way for localization in sparse packages and we rely on this key for localized title of the menu.

deepak1556 commented on build/win32/code.iss:
```diff
@@ -1376,6 +1381,11 @@ begin
 
+function IsWindows11OrLater(): Boolean;
+begin
+  Result := (GetWindowsVersion >= $0A0055F0);
```
Please refer to `Pascal Scripting: GetWindowsVersion` section for how the version bits are aligned https://jrsoftware.org/ishelp/

deepak1556 commented:
**Status:**

Highlighting some interesting corner cases when creating sparse packages with Chromium runtime,

- **Fixed** Installing the package with `ExternalLocation` set to the root of the VSCode installation folder leads to a crash on startup when application is installed under `%LOCALAPPDATA/Programs` (User setup)
   - **Problem** Child process that run with sandbox gets aborted by the runtime as soon as the sandbox phase was initialized after the process launch, currently in VSCode case it would be GPU process and some workbench processes (Process Explorer, DevTools, Issue Reporter, OOPIF). Application launches with Medium IL but the sandbox will change IL to Low IL (GPU) and Untrusted IL (Renderer). Turns out the Appx installation adds two new SID (S-1-15-2-xxx and S-1-15-3-xxx) to the folder marked as `ExternalLocation` and changes the IL of the folder https://devblogs.microsoft.com/oldnewthing/20220502-00/?p=106550. Based on the exit code of the crashing process, it seemed that process were unable to load DLLs. Both GPU and Renderer load certain DLLs present in the VSCode installation folder as part of the initialization phase and the lower IL of these processes causes the abort.
   - **Solution** Install the sparse package resources from a child folder of the VSCode Installation path so that the root folder permissions remain unchanged.

- **Fixed** User setup fails to persist data to the default user data path `%APPDATA%` when launched via the context menu
   - **Problem** Capability `unvirtualizedResources` added to the package enables filesystem virtualization by default https://docs.microsoft.com/en-us/windows/uwp/packaging/app-capability-declarations
   - **Solution** Disable `FileSystemWriteVirtualization`
 
- **Fixed** Localization of context menu title using a resources file does not work out of box https://docs.microsoft.com/en-us/windows/uwp/app-resources/localize-strings-ui-manifest#loading-strings-in-non-packaged-applications
   - **Problem** `ResourceLoader::GetForViewIndependentUse` does not pick up the resources file when called from within the sparse DLL. Most Electron apps, don't have control over the runtime code and it is not possible to integrate the resource file for the main application.
   - **Workaround** Read from the localized strings in the registry that were setup as part of the old context menu.

- **Pending** Integrate sparse package with the Inno updater flow
   - **Problem** During update inno will wait for the application to exit gracefully then replace the files of the installation folder and trigger a restart. Context menu integration loads a DLL as part of the windows explorer which needs to get unloaded (via Remove-AppxPackage cmdlet) before the files are replaced by inno. This requires we make the updater code aware of the sparse package details which is not ideal as we have been keeping the updater isolated from the application logic https://github.com/Microsoft/inno-updater.

deepak1556 commented:
@joaomoreno Selfhosting build available at
![selfhost-build](https://user-images.githubusercontent.com/964386/177818836-1d9c5fc7-986f-4a8c-8446-47c5c2e2c963.png)

Curious-r commented:
> Next Steps to land this for preview with Insiders:
> 
> * [ ]  Make inno-updater skip appx package folder during update
> * [ ]  Separate the context menu into new tasks and isolate them for Insiders flavor
> * [ ]  Highlight reinstallation to udpate the appx package in release notes.

Exciting!

deepak1556 commented:
New insiders build : https://monacotools.visualstudio.com/DefaultCollection/Monaco/_build/results?buildId=178917&view=results

joaomoreno commented:
New one with updated inno-updater: https://monacotools.visualstudio.com/DefaultCollection/Monaco/_build/results?buildId=178931&view=results

joaomoreno commented:
@deepak1556 Getting this when running the setup, possibly related to https://github.com/microsoft/vscode/pull/151186/files#diff-8dd373ae4b3c58948172ac6408e9e57ff2451265eca1b3b3c27682b79869bfceR1387

![image](https://user-images.githubusercontent.com/22350/180233945-04ec3c80-c8a7-4ae7-8998-3e6bea499cdd.png)

deepak1556 commented:
Thanks for checking! That condition should not be an issue since it worked in previous builds, I think it is from this newly added condition https://github.com/microsoft/vscode/pull/151186/files#diff-8dd373ae4b3c58948172ac6408e9e57ff2451265eca1b3b3c27682b79869bfceR1416.

deepak1556 commented:
New build https://monacotools.visualstudio.com/DefaultCollection/Monaco/_build/results?buildId=178950&view=results

deepak1556 commented:
@joaomoreno the builds from the last run are good for testing, PTAL. Thanks!

deepak1556 commented:
Build for testing: https://monacotools.visualstudio.com/Monaco/_build/results?buildId=189029&view=results

deepak1556 commented:
New run: https://monacotools.visualstudio.com/Monaco/_build/results?buildId=189370&view=results

joaomoreno commented:
Here are some testing findings.

### Corrupted menu entry

1. Install user setup without the option to add to context menu
2. Right click a folder
 
🐛 Menu entry appears garbled

![Screenshot 2022-10-18 202608](https://user-images.githubusercontent.com/22350/196513893-90c81d3f-00b1-45d6-bd30-3c78ea2805f7.png)

### Missing entry during update

1. Install user setup with the option to add to context menu
2. Run Code
3. Run `Developer: Apply Update`, select the same user setup package
4. Wait for the gear badge to show an update is ready
5. Right click a directory in the File Explorer

🐛 The menu entry is missing

### ⚠️ Broken background update

(continue from previous scenario)

6. Click the gear icon and choose `Restart to Update`

🐛 Besides taking too long, there's now an error. Pressing Retry does not work, just shows the message again. Pressing Cancel is the only option, and this will brick the installation.

![image](https://user-images.githubusercontent.com/22350/196516485-1fc47e4e-fcee-4822-a5e7-7b66e128ec4c.png)

deepak1556 commented:
New Build: https://monacotools.visualstudio.com/Monaco/_build/results?buildId=189784&view=results

Addresses the issues outlined in https://github.com/microsoft/vscode/pull/151186#issuecomment-1282854207

joaomoreno commented:
A few more issues:

### Garbled menu entry in system setup

1. Install system setup

🐛 The menu item is garbled

![image](https://user-images.githubusercontent.com/22350/196947081-c20b4bf8-c893-4e78-8247-217c675c8ba3.png)

### Context menu entry not removed

1. Install VS Code with context menu enabled
3. Without uninstalling, reinstall VS Code and this time disable the context menu

🐛 The context menu is still there.

### Double menu entry in classic menu after update

> I'm fine if we don't fix this, given that it only surfaces in the classic menu and only applies to updates. A fresh installation doesn't hit this issue.

1. Install previous VS Code version, enable context menu
2. Use the `Developer: Apply Update` to update to the new version

:bug: The classic menu will contain both action items

![image](https://user-images.githubusercontent.com/22350/196944631-3ab1310c-bc69-44fd-b3de-9f20f351d660.png)

### No menu entry in arm64

1. Install VS Code with context menu enabled, in Windows for ARM64

:bug: No menu entry is present

The AppX seems registered though:

![image](https://user-images.githubusercontent.com/22350/196948520-ec90f72c-3980-4ea7-94b3-5ccbfb3901ca.png)

deepak1556 commented:
New Build: https://monacotools.visualstudio.com/Monaco/_build/results?buildId=189968&view=results

Addresses the following issues from https://github.com/microsoft/vscode/pull/151186#issuecomment-1285473301

- Garbled menu entry in system setup
- Context menu entry not removed
- Double menu entry in classic menu after update

I haven't setup Windows on arm on my new device yet, will verify the last pending issue after that.

joaomoreno commented:
Getting there! This whole infrastructure is soooo fragile.

More findings:

### Error when reinstalling

1. Install VS Code with context menu entry
2. Without uninstalling, reinstall VS Code

🐛 There's an error about `COM Surrogate`:

- Choosing _Automatically close the applications_ doesn't work
- Choosing _Do not close the applications_ seems to work actually!

![image](https://user-images.githubusercontent.com/22350/197194455-c50382cc-39c3-4be8-8db1-d0bcd51a049e.png)

### Rolling back doesn't remove context menu entry

1. Install VS code with context menu entry
2. Use the `Developer: Apply Update` to update to an old version

🐛 Context menu entry is still there

deepak1556 commented:
`DLLHost.exe` aka [COM surrogate](https://devblogs.microsoft.com/oldnewthing/20090212-00/?p=19173) holds the DLL after context menu is closed for a certain short period of time, I think this has to do with how long the Explorer references the COM object but from my testing it is not an unacceptable long period. But during this period if we perform the update step then we will run into the above error. This is an edge case we cannot solve from our DLL code or the updater. I think this case would be rarely hit by an user, I plan to ignore this issue for now and revisit based on insiders testing.

As for the rollback scenario, we don't plan to support it since it requires changes in our updater to know about the appx package. I am going to merge this PR for insiders testing and follow-up on the arm64 issue.

@joaomoreno thanks a lot for the extensive testing! ❤️

joaomoreno commented on build/azure-pipelines/win32/product-build-win32.yml:
```diff
@@ -271,6 +282,7 @@ steps:
         . build/azure-pipelines/win32/exec.ps1
         $ErrorActionPreference = "Stop"
         exec { node build\azure-pipelines\common\sign $env:EsrpCliDllPath windows $(ESRP-PKI) $(esrp-aad-username) $(esrp-aad-password) $(CodeSigningFolderPath) '*.dll,*.exe,*.node' }
+        exec { node build\azure-pipelines\common\sign $env:EsrpCliDllPath windows-appx $(ESRP-PKI) $(esrp-aad-username) $(esrp-aad-password) $(CodeSigningFolderPath) '*.appx' }
```
@deepak1556 Just noticed this: do we need an `if insiders` here as well?

joaomoreno commented on build/win32/code.iss:
```diff
@@ -34,6 +34,7 @@ ShowLanguageDialog=auto
 ArchitecturesAllowed={#ArchitecturesAllowed}
 ArchitecturesInstallIn64BitMode={#ArchitecturesInstallIn64BitMode}
 WizardStyle=modern
+SetupLogging=yes
```
@deepak1556 We can now remove this.

joaomoreno commented on build/win32/code.iss:
```diff
@@ -86,8 +87,9 @@ Name: "runcode"; Description: "{cm:RunAfter,{#NameShort}}"; GroupDescription: "{
-Source: "*"; Excludes: "\CodeSignSummary*.md,\tools,\tools\*,\resources\app\product.json"; DestDir: "{code:GetDestDir}"; Flags: ignoreversion recursesubdirs createallsubdirs
+Source: "*"; Excludes: "\CodeSignSummary*.md,\tools,\tools\*,\appx,\appx\*,\resources\app\product.json"; DestDir: "{code:GetDestDir}"; Flags: ignoreversion recursesubdirs createallsubdirs
 Source: "tools\*"; DestDir: "{app}\tools"; Flags: ignoreversion
+Source: "appx\*"; DestDir: "{app}\appx"; BeforeInstall: RemoveAppxPackage; AfterInstall: AddAppxPackage; Flags: ignoreversion
```
@deepak1556 Do we need an `if insiders` here?

deepak1556 commented:
@joaomoreno addressed in https://github.com/microsoft/vscode/pull/164411
